aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/SyncProcessor.java

```diff
@@ -235,10 +229,28 @@ private SyncTime filterOutOldSyncTimes(SyncTime lastSyncTime) throws DataStoreEx
                         processor.onComplete();
                     }
                 })
-                // Flatten the PaginatedResult objects into a stream of ModelWithMetadata objects.
-                .concatMapIterable(PaginatedResult::getItems)
-                // Stop after fetching the maximum configured records to sync.
-                .take(dataStoreConfigurationProvider.getConfiguration().getSyncMaxRecords());
+                // If it's a SerializedModel, add the ModelSchema, since it isn't added during deserialization.
+                .map(paginatedResult -> Flowable.fromIterable(paginatedResult)
+                        .map(modelWithMetadata -> hydrateSchemaIfNeeded(modelWithMetadata, schema))
+                        .toList()
+                        .blockingGet()
+                )
+                .takeUntil(items -> recordsFetched.accumulateAndGet(items.size(), Integer::sum) >= syncMaxRecords);
```

Comment:
Were we not actually enforcing this logic anywhere, before?

Reply:
We were.  Previously, it was `.take(syncMaxRecords)`, because it was flattened into a `Flowable<ModelWithMetadata<T>>` in this method, first.  Since it's returning a `Flowable<List<ModelWithMetadata<T>>>` now so that we can do batch saves in the future, I had to use `takeUntil` and increment a counter manually.